### PR TITLE
Use bazel to build/push kubemark image

### PR DESCRIPTION
**Pull request by krzyzacy**
try to get some proof of concept, kubemark image is probably simple enough to get converted to bazel. (me bazel noob still trying it out locally)

cc @BenTheElder @ixdy @shyamjvs 

/release-note-none

### CODE REVIEW

**Comment by cblecker** [comment]
/unassign
bazel isn't my area of expertise :)

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/test-infra/issues/4947

File: build/root/BUILD.root [hunk 1]
```diff
@@ -65,6 +65,16 @@ filegroup(
...
+        "//kubernetes/cluster:all-srcs",
+        "//kubernetes/docs:all-srcs",
+        "//kubernetes/examples:all-srcs",
+        "//kubernetes/federation/cluster:all-srcs",
+        "//kubernetes/federation/develop:all-srcs",
+        "//kubernetes/hack:all-srcs",
+        "//kubernetes/test/e2e/testing-manifests:all-srcs",
+        "//kubernetes/test/images:all-srcs",
+        "//kubernetes/test/kubemark:all-srcs",
+        "//kubernetes/third_party/htpasswd:all-srcs",
```
Comment (krzyzacy) [inline build/root/BUILD.root]: hummm, those were wrote by `./hack/update-bazel.sh`

Comment (ixdy) [reply build/root/BUILD.root]: you probably have a kubernetes/ dir in the root of your repo, probably due to extracting a `kubernetes.tar.gz` there. remove it and re-run `update-bazel.sh`.

File: cluster/images/kubemark/BUILD [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
+
+load("@io_bazel_rules_docker//docker:docker.bzl", "docker_build", "docker_push")
+
+docker_build(
+    name = "image",
+    base = "@debian//image",
```
Comment (ixdy) [inline cluster/images/kubemark/BUILD]: I don't think kubemark needs debian anymore; we might be able to get away with busybox or scratch now.

File: cluster/images/kubemark/BUILD [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    entrypoint = ["/kubemark"],
+    files = [":kubemark"],
+)
+
+docker_push(
+    name = "push",
+    image = ":image",
+    registry = "gcr.io",
+    repository = "google_containers/kubemark",
```
Comment (ixdy) [inline cluster/images/kubemark/BUILD]: prefer "google-containers" to "google_containers".

File: cluster/images/kubemark/BUILD [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+licenses(["notice"])
+
+load("@io_bazel_rules_docker//docker:docker.bzl", "docker_build", "docker_push")
+
+docker_build(
+    name = "image",
+    base = "@debian//image",
+    entrypoint = ["/kubemark"],
+    files = [":kubemark"],
```
Comment (ixdy) [inline cluster/images/kubemark/BUILD]: this should depend on the kubemark binary properly, unlike the way the Dockerfile/Makefile is currently written.

Comment (krzyzacy) [reply cluster/images/kubemark/BUILD]: you mean have another bazel rule to build the kubemark binary?

**Comment by krzyzacy** [comment]
getting some weird import python error right now

```
ERROR: /usr/local/google/home/senlu/work/src/k8s.io/kubernetes/cluster/images/kubemark/BUILD:7:1: no such package '@debian//image': Pull command failed: Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/__main__.py", line 93, in <module>
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/__main__.py", line 86, in main
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/containerregistry/client/v2_2/v2_compat_.py", line 106, in __init__
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/containerregistry/client/v2_2/v2_compat_.py", line 110, in _ProcessImage
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/containerregistry/client/v2/docker_image_.py", line 177, in manifest
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/containerregistry/client/v2/docker_digest_.py", line 32, in SignedManifestToSHA256
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par/containerregistry/client/v2/util_.py", line 96, in DetachSignatures
KeyError: 'signatures'
 (/usr/bin/python /usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/puller/file/puller.par --directory /usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/external/debian/image --name index.docker.io/library/debian@sha256:f51cf81db2de8b5e9585300f655549812cdb27c56f8bfb992b8b706378cd517d) and referenced by '//cluster/images/kubemark:image'
ERROR: Analysis of target '//cluster/images/kubemark:image' failed; build aborted
INFO: Elapsed time: 4.790s
FAILED: Build did NOT complete successfully (7 packages loaded)
ERROR: Build failed. Not running target
```

File: cluster/images/kubemark/BUILD (hunk 4 above)
Comment (ixdy) [reply cluster/images/kubemark/BUILD]: well, the rule should already exist. you probably want `//cmd/kubemark` here.

**Comment by krzyzacy** [comment]
Local build now works, but got some other connection error(?) upon pushing. (Also added some conditional env to test bazel build/upload in start-kubemark.sh)

```
senlu@senlu:~/work/src/k8s.io/kubernetes$ bazel run //cluster/images/kubemark:push
INFO: Analysed target //cluster/images/kubemark:push (1 packages loaded).
INFO: Found 1 target...
Target //cluster/images/kubemark:push up-to-date:
  bazel-bin/cluster/images/kubemark/push
INFO: Elapsed time: 0.849s, Critical Path: 0.17s
INFO: Build completed successfully, 2 total actions

INFO: Running command line: bazel-bin/cluster/images/kubemark/push
ERROR:root:Error during upload of: gcr.io/google-containers/kubemark:latest
Traceback (most recent call last):
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/__main__/../containerregistry/tools/fast_pusher_.py", line 141, in <module>
    main()
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/__main__/../containerregistry/tools/fast_pusher_.py", line 132, in main
    session.upload(v2_2_img)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/client/v2_2/docker_session_.py", line 314, in upload
    future.result()
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/concurrent/futures/_base.py", line 398, in result
    return self.__get_result()
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/client/v2_2/docker_session_.py", line 275, in _upload_one
    self._put_blob(image, digest)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/client/v2_2/docker_session_.py", line 201, in _put_blob
    self._patch_upload(image, digest)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/client/v2_2/docker_session_.py", line 167, in _patch_upload
    accepted_codes=[httplib.NO_CONTENT, httplib.ACCEPTED, httplib.CREATED])
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/client/v2_2/docker_http_.py", line 353, in Request
    url, method, body=body, headers=headers)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/containerregistry/transport/transport_pool_.py", line 62, in request
    return transport.request(*args, **kwargs)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/httplib2/__init__.py", line 1659, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/httplib2/__init__.py", line 1399, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "/usr/local/google/home/senlu/.cache/bazel/_bazel_senlu/8bed9ea5a35c4e84a8a5913ca5672618/execroot/__main__/bazel-out/local-fastbuild/bin/cluster/images/kubemark/push.runfiles/httplib2/__init__.py", line 1382, in _conn_request
    content = response.read()
  File "/usr/lib/python2.7/httplib.py", line 578, in read
    return self._read_chunked(amt)
  File "/usr/lib/python2.7/httplib.py", line 632, in _read_chunked
    raise IncompleteRead(''.join(value))
httplib.IncompleteRead: IncompleteRead(73 bytes read)
ERROR: Non-zero return code '1' from command: Process exited with status 1
```

File: test/kubemark/start-kubemark.sh [hunk 5]
```diff
@@ -214,6 +214,25 @@ function create-and-upload-hollow-node-image {
   echo "Created and uploaded the kubemark hollow-node image to docker registry."
 }
 
+# Use bazel rule to create a docker image for hollow-node and upload
+# it to the appropriate docker container registry for the cloud provider.
+function create-and-upload-hollow-node-image-bazel {
+  RETRIES=3
+  for attempt in $(seq 1 ${RETRIES}); do
+    if ! bazel run //cluster/image/kubemark:push; then
```
Comment (shyamjvs) [inline test/kubemark/start-kubemark.sh]: s/image/images

File: test/kubemark/start-kubemark.sh [hunk 6]
```diff
@@ -214,6 +214,25 @@ function create-and-upload-hollow-node-image {
   echo "Created and uploaded the kubemark hollow-node image to docker registry."
 }
 
+# Use bazel rule to create a docker image for hollow-node and upload
+# it to the appropriate docker container registry for the cloud provider.
+function create-and-upload-hollow-node-image-bazel {
```
Comment (shyamjvs) [inline test/kubemark/start-kubemark.sh]: Most of this function's logic is similar to `create-and-upload-hollow-node-image`. The main difference is we're calling bazel here and make there. Could you change it to switch b/w both modes (based on a flag or sth)>

**Comment by shyamjvs** [comment]
Couple of comments and lgtm. Thanks for the PR.
Also please verify that it works.

File: test/kubemark/start-kubemark.sh (hunk 6 above)
Comment (krzyzacy) [reply test/kubemark/start-kubemark.sh]: it looks for `KUBEMARK_BAZEL_BUILD`, if that's what you meant?

Comment (shyamjvs) [reply test/kubemark/start-kubemark.sh]: Yup, saw that.. But you still don't seem to need 2 functions pretty much replicating the same logic.

Comment (krzyzacy) [reply test/kubemark/start-kubemark.sh]: My intend is to swap out the entire function later on, seems this is clearer than embed more if/else into the existing function?

**Comment by krzyzacy** [comment]
okay the bazel rule worked locally now :-)

File: test/kubemark/start-kubemark.sh (hunk 6 above)
Comment (shyamjvs) [reply test/kubemark/start-kubemark.sh]: Sgtm

**Comment by krzyzacy** [comment]
/assign @mikedanese 
for approval

File: cluster/images/kubemark/BUILD [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
```
Comment (ixdy) [inline cluster/images/kubemark/BUILD]: don't need this anymore

File: cluster/images/kubemark/BUILD [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+docker_push(
+    name = "push",
+    image = ":image",
+    registry = "gcr.io",
+    repository = "$(PROJECT)/kubemark",
+    tag = "latest",
+
+    # Trigger stamping.
```
Comment (ixdy) [inline cluster/images/kubemark/BUILD]: don't need this comment

**Review by ixdy — COMMENTED** [review]
basically LG, a few minor comments

File: cluster/images/kubemark/BUILD (hunk 7 above)
Comment (krzyzacy) [reply cluster/images/kubemark/BUILD]: done (thought that was from update-bazel.sh)

File: cluster/images/kubemark/BUILD (hunk 8 above)
Comment (krzyzacy) [reply cluster/images/kubemark/BUILD]: deleted

**Comment by krzyzacy** [comment]
(can someone stamp again?)

**Comment by krzyzacy** [comment]
thanks!

**Comment by ixdy** [comment]
I think this broke `make bazel-build`:
```console
$ make bazel-build
...................
ERROR: [elided]/src/k8s.io/kubernetes/cluster/images/kubemark/BUILD:12:1: in repository attribute of container_push rule //cluster/images/kubemark:push: $(PROJECT) not defined. Since this rule was created by the macro 'docker_push', the error might have been caused by the macro implementation in [elided]/src/k8s.io/kubernetes/cluster/images/kubemark/BUILD:12:1
ERROR: Analysis of target '//cluster/images/kubemark:push' failed; build aborted
INFO: Elapsed time: 30.148s
FAILED: Build did NOT complete successfully (2816 packages loaded)
make: *** [bazel-build] Error 1
```

**Comment by krzyzacy** [comment]
hummmmmm, a separate issue - k8s tarball does not include bazel workspace

**Comment by ixdy** [comment]
well, there are probably multiple things broken.

the issue I'm noting here is that `bazel build //...` is trying to build `'//cluster/images/kubemark:push` and failing.

**Comment by krzyzacy** [comment]
(fixed by https://github.com/kubernetes/kubernetes/pull/53915)

**Comment by BenTheElder** [comment]
> (fixed by #53915)

(which is merged now 🚒)

**Pull request merged**